# Desafío 1

__Importo librerías y el dataset__

In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', 500)
from scipy import stats
import seaborn as sns
import matplotlib as plt
from scipy.stats import norm
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
scaler=StandardScaler()
import plotly.plotly as py
import plotly.graph_objs as go
from sklearn.linear_model import ElasticNet, Lasso,  BayesianRidge, LassoLarsIC
from sklearn.ensemble import RandomForestRegressor,  GradientBoostingRegressor
from sklearn.kernel_ridge import KernelRidge
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import RobustScaler
from sklearn.base import BaseEstimator, TransformerMixin, RegressorMixin, clone
from sklearn.model_selection import KFold, cross_val_score, train_test_split
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd
from scipy import stats
import seaborn as sns
from sklearn import metrics
from sklearn import datasets
from sklearn.model_selection import train_test_split, KFold, cross_val_score,StratifiedKFold
from sklearn.linear_model import LinearRegression, Lasso, LassoCV, Ridge, RidgeCV
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import ElasticNet, Lasso,  BayesianRidge, LassoLarsIC,LogisticRegression
from sklearn.ensemble import RandomForestRegressor,  GradientBoostingRegressor
from sklearn.kernel_ridge import KernelRidge
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import RobustScaler
from sklearn.base import BaseEstimator, TransformerMixin, RegressorMixin, clone
from sklearn.model_selection import KFold, cross_val_score, train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor
import xgboost as xgb
from xgboost import XGBRegressor
import lightgbm as lgb
!pip install vecstack
from vecstack import StackingTransformer, stacking
from scipy.stats import norm
from sklearn.preprocessing import StandardScaler
from scipy import stats
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline
from scipy.stats import skew
from scipy.special import boxcox1p
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
rmse = lambda y, y_pred: np.sqrt(mean_squared_error(y, y_pred))

/anaconda3/lib/python3.7/site-packages/lightgbm/__init__.py:46: UserWarning:

Starting from version 2.2.1, the library file in distribution wheels for macOS is built by the Apple Clang (Xcode_8.3.1) compiler.
This means that in case of installing LightGBM from PyPI via the ``pip install lightgbm`` command, you don't need to install the gcc compiler anymore.
Instead of that, you need to install the OpenMP library, which is required for running LightGBM on the system with the Apple Clang compiler.
You can install the OpenMP library by the following command: ``brew install libomp``.



In [2]:
df=pd.read_csv("properatti.csv", index_col=0)

In [3]:
df.head()

operation property_type place_name  \
0      sell            PH  Mataderos   
1      sell     apartment   La Plata   
2      sell     apartment  Mataderos   
3      sell            PH    Liniers   
4      sell     apartment     Centro   

                             place_with_parent_names country_name  \
0              |Argentina|Capital Federal|Mataderos|    Argentina   
1        |Argentina|Bs.As. G.B.A. Zona Sur|La Plata|    Argentina   
2              |Argentina|Capital Federal|Mataderos|    Argentina   
3                |Argentina|Capital Federal|Liniers|    Argentina   
4  |Argentina|Buenos Aires Costa Atlántica|Mar de...    Argentina   

                     state_name  geonames_id                  lat-lon  \
0               Capital Federal    3430787.0  -34.6618237,-58.5088387   
1        Bs.As. G.B.A. Zona Sur    3432039.0  -34.9038831,-57.9643295   
2               Capital Federal    3430787.0  -34.6522615,-58.5229825   
3               Capital Federal    3431333.0  -34.6477969,-58.5164244   
4  Buenos Aires Costa Atlántica    3435548.0  -38.0026256,-57.5494468   

         lat        lon     price currency  price_aprox_local_currency  \
0 -34.661824 -58.508839   62000.0      USD                   1093959.0   
1 -34.903883 -57.964330  150000.0      USD                   2646675.0   
2 -34.652262 -58.522982   72000.0      USD                   1270404.0   
3 -34.647797 -58.516424   95000.0      USD                   1676227.5   
4 -38.002626 -57.549447   64000.0      USD                   1129248.0   

   price_aprox_usd  surface_total_in_m2  surface_covered_in_m2  \
0          62000.0                 55.0                   40.0   
1         150000.0                  NaN                    NaN   
2          72000.0                 55.0                   55.0   
3          95000.0                  NaN                    NaN   
4          64000.0                 35.0                   35.0   

   price_usd_per_m2  price_per_m2  floor  rooms  expenses  \
0       1127.272727   1550.000000    NaN    NaN       NaN   
1               NaN           NaN    NaN    NaN       NaN   
2       1309.090909   1309.090909    NaN    NaN       NaN   
3               NaN           NaN    NaN    NaN       NaN   
4       1828.571429   1828.571429    NaN    NaN       NaN   

                                       properati_url  \
0  http://www.properati.com.ar/15bo8_venta_ph_mat...   
1  http://www.properati.com.ar/15bob_venta_depart...   
2  http://www.properati.com.ar/15bod_venta_depart...   
3  http://www.properati.com.ar/15boh_venta_ph_lin...   
4  http://www.properati.com.ar/15bok_venta_depart...   

                                         description  \
0  2 AMBIENTES TIPO CASA PLANTA BAJA POR PASILLO,...   
1  Venta de departamento en décimo piso al frente...   
2  2 AMBIENTES 3ER PISO LATERAL LIVING COMEDOR AM...   
3  PH 3 ambientes con patio. Hay 3 deptos en lote...   
4  DEPARTAMENTO CON FANTÁSTICA ILUMINACIÓN NATURA...   

                                               title  \
0                 2 AMB TIPO CASA SIN EXPENSAS EN PB   
1  VENTA Depto 2 dorm. a estrenar 7 e/ 36 y 37   ...   
2           2 AMB 3ER PISO CON ASCENSOR APTO CREDITO   
3                       PH 3 amb. cfte. reciclado      
4  DEPTO 2 AMB AL CONTRAFRENTE ZONA CENTRO/PLAZA ...   

                                     image_thumbnail  
0  https://thumbs4.properati.com/8/BluUYiHJLhgIIK...  
1  https://thumbs4.properati.com/7/ikpVBu2ztHA7jv...  
2  https://thumbs4.properati.com/5/SXKr34F_IwG3W_...  
3  https://thumbs4.properati.com/3/DgIfX-85Mog5SP...  
4  https://thumbs4.properati.com/5/xrRqlNcSI_vs-f...

__Tomo el tipo de cambio en el caso de necesitarlo más adelante__

In [4]:
tc = round(np.mean(df.price_aprox_local_currency / df.price_aprox_usd),4)

__Chequeo de variables categóricas__

In [5]:
df["operation"].value_counts()

sell    121220
Name: operation, dtype: int64

Como todas las operacines son de venta, procedo a eliminar la columna ya que no me aporta información diferencial entre filas.

In [6]:
df.drop(columns=['operation'], inplace=True)

In [7]:
df["property_type"].value_counts()

apartment    71065
house        40268
PH            5751
store         4136
Name: property_type, dtype: int64

In [8]:
df["place_name"].value_counts()

Córdoba                        9254
Rosario                        8504
Mar del Plata                  6534
Tigre                          3324
Nordelta                       3315
Belgrano                       2992
Palermo                        2885
La Plata                       2534
Caballito                      2273
Pilar                          2228
Villa Urquiza                  1632
Olivos                         1590
Recoleta                       1547
Lomas de Zamora                1529
Flores                         1354
Ituzaingó                      1340
Villa Crespo                   1331
Capital Federal                1297
San Telmo                      1216
Morón                          1209
Almagro                        1165
Barrio Norte                   1140
Ramos Mejía                    1127
Martínez                       1058
Lanús                          1048
Escobar                        1037
Banfield                       1007
San Isidro                  

In [9]:
df["state_name"].value_counts()

Capital Federal                 32316
Bs.As. G.B.A. Zona Norte        25560
Bs.As. G.B.A. Zona Sur          13952
Córdoba                         12069
Santa Fe                        10172
Buenos Aires Costa Atlántica    10006
Bs.As. G.B.A. Zona Oeste         9322
Buenos Aires Interior            2291
Río Negro                         808
Neuquén                           733
Mendoza                           681
Tucumán                           674
Corrientes                        583
Misiones                          464
Entre Ríos                        369
Salta                             278
Chubut                            259
San Luis                          252
La Pampa                          157
Formosa                            65
Chaco                              57
San Juan                           40
Tierra Del Fuego                   31
Catamarca                          27
Jujuy                              26
Santa Cruz                         20
Santiago Del

In [10]:
df['currency'].value_counts()

USD    87587
ARS    13219
PEN        2
UYU        1
Name: currency, dtype: int64

## Precios en USD

__Como hay 3 valores que no están ni en pesos ni en dólares se procede a eliminarlos__

In [11]:
df = df[df.currency != 'UYU']
df = df[df.currency != 'PEN']

In [12]:
df.isnull().sum()

property_type                      0
place_name                        23
place_with_parent_names            0
country_name                       0
state_name                         0
geonames_id                    18717
lat-lon                        51547
lat                            51547
lon                            51547
price                          20410
currency                       20411
price_aprox_local_currency     20410
price_aprox_usd                20410
surface_total_in_m2            39325
surface_covered_in_m2          19904
price_usd_per_m2               52600
price_per_m2                   33559
floor                         113318
rooms                          73827
expenses                      106955
properati_url                      0
description                        2
title                              0
image_thumbnail                 3112
dtype: int64

__Chequeo que no haya propiedades en dólares cuyo precio sea mayor al precio en pesos__

In [13]:
len(df[(df['currency'] == 'USD') & (df['price'] > df['price_aprox_local_currency'])])

0

Como da 0 sabemos que no hay precios en dólares mayores a los valores en pesos

__Chequeo que no haya propiedades en pesos cuyo precio sea menor al precio en dolares__

In [14]:
len(df[(df['currency'] == 'ARS') & (df['price'] < df['price_aprox_usd'])])

0

Como da 0 sabemos que no hay precios en pesos menores a los precios en USD

__Chequeo que el precio en dólares sea igual al precio en moneda extranjera (USD)__

In [15]:
len(df[(df['currency'] == 'USD') & (df['price'] != df['price_aprox_usd'])])

0

Como da 0 sabemos que no hay diferencia entre los precios en USD

__Chequeo que el precio en ARS sea igual al precio local (ARS)__

In [16]:
len(df[(df['currency'] == 'ARS') & (df['price'] != df['price_aprox_local_currency'])])

13219

In [17]:
len(df[df['currency'] == 'ARS'])

13219

Como se puede ver hay diferencia en todos los precios en ARS vs precio Local, se decide tomar la columna de price aprox USD como precio

In [18]:
df['precioReal1'] = df['price_aprox_usd']

__Chequeo cuantas propiedades no tienen precio asignado__

In [19]:
df.precioReal1.isnull().sum()

20410

__Si es nulo el precio, verifico que no haya alguna otra columna de precio que tenga valor__

In [20]:
len(df[(df.precioReal1.isnull()) & (~df.price_aprox_local_currency.isnull() | 
                                    ~df.price_aprox_usd.isnull() | 
                                    ~df.price.isnull())])

0

Comprobamos que no hay ninguna referencia de precio para reemplazar directamente

__De los registros que no tienen precio, ver cual tiene precio por m2__

In [21]:
len(df[(df.precioReal1.isnull()) & (~df.price_usd_per_m2.isnull() | ~df.price_per_m2.isnull())])

0

No hay registros del que se pueda inferir el precio por el precio por m2

__Al ver el título y la descripción de las propiedades, se puede ver que hay algunos que poseen el valor de la propiedad, con lo cual hay que extraer la información.__

In [22]:
df['regex_titulo'] = df['title'].str.extract('(U[$S]D [0-9]?[0-9]?[0-9]?[\.\s]?[0-9]?[0-9]?[0-9]?[\.\s]?[0-9]?[0-9]?[0-9]?)', expand=False)
len(df[df['regex_titulo'].notnull()])

2539

Hay 2539 valores a los cuales le podemos extraer el valor en dolares de la propiedad del titulo

__Elimino el signo dólares y conservo los valores de casas que valgan más de 10 mil USD__

In [23]:
df['regex_titulo'] = df['regex_titulo'].str.extract('([0-9]?[0-9]?[0-9]?[\.\s]?[0-9]?[0-9][0-9][\.\s]?[0-9][0-9][0-9])',
                                                    expand=False)
len(df[df['regex_titulo'].notnull()])

2507

__Elimino los puntos intermedios para luego convertir a float__

In [24]:
df.regex_titulo.replace(to_replace='\.',value='',inplace=True, regex=True)
len(df[df.regex_titulo.notnull()])

2507

In [25]:
df.regex_titulo = df.regex_titulo.astype(float)

Hacemos un chequeo rápido de que la conversión fue efectiva

In [26]:
df.regex_titulo.dtype

dtype('float64')

__Para todos los precios reales que son NaN, tomo los valores de regex_titulo__

In [27]:
df.loc[df.precioReal1.notnull(),'precioReal2'] = df.loc[df.precioReal1.notnull(),'precioReal1']
df.loc[((df.precioReal1.isnull()) & (df.regex_titulo.notnull())),'precioReal2'] = df.loc[((df.precioReal1.isnull()) & (df.regex_titulo.notnull())),'regex_titulo']

In [28]:
df.precioReal2.isnull().sum()

18010

Chequeo que los valores se hayan reemplazado

In [29]:
df[(df.precioReal1.isnull() & df.regex_titulo.notnull())]['precioReal2'].isnull().sum()

0

__Así como en el título, hay descripciones que poseen valores de propiedad en dolares por lo que los extraemos del mismo__

In [30]:
df['regex_desc'] = df['description'].str.extract('(U[$S]D [0-9]?[0-9]?[0-9]?[\.\s]?[0-9]?[0-9]?[0-9]?[\.\s]?[0-9]?[0-9]?[0-9]?)',
                                                 expand=False)
len(df[df['regex_desc'].notnull()])

5882

Hay 5882 valores a los cuales le podemos extraer el valor en dolares de la propiedad de descripcion

__Elimino el signo dolares y me quedo con los valores de casas que valgan más de 10.000 dolares__

In [31]:
df['regex_desc'] = df['regex_desc'].str.extract('([0-9]?[0-9]?[0-9]?[\.\s]?[0-9]?[0-9][0-9][\.\s]?[0-9][0-9][0-9])',
                                                expand=False)
len(df[df['regex_desc'].notnull()])

5273

__Elimino los puntos intermedios para luego convertir a float__

In [32]:
df.regex_desc.replace(to_replace='\.',value='',inplace=True, regex=True)
len(df[df.regex_desc.notnull()])

5273

In [33]:
df.regex_desc = df.regex_desc.astype(float)

Chequeo conversión

In [34]:
df.regex_desc.dtype

dtype('float64')

__Para todos los preciosReales2 que son NaN, tomo los valores de regex_desc__

In [35]:
df.loc[df.precioReal2.notnull(),'precio_usd'] = df.loc[df.precioReal2.notnull(),'precioReal2']
df.loc[((df.precioReal2.isnull()) & 
        (df.regex_desc.notnull())),'precio_usd'] = df.loc[((df.precioReal2.isnull()) & 
                                                           (df.regex_desc.notnull())),'regex_desc']

Chequeo que los valores se hayan reemplazado

In [36]:
df[(df.precioReal2.isnull() & df.regex_desc.notnull())]['precio_usd'].isnull().sum()

0

__Vemos la cantidad de valores que recupere con la descripcion__

In [37]:
df.precioReal2.isnull().sum() - df.precio_usd.isnull().sum()

20

__Ahora que ya tengo todos los precios en dolares directos, procedo a chequear si puedo salvar registros con el precio por metro cuadrado__

In [38]:
len(df[df.precio_usd.isnull() & (df.price_usd_per_m2.notnull() | df.price_per_m2.notnull())])

0

Como es 0, elimino las filas ya que no puedo recuperar el precio

In [39]:
df.dropna(subset=['precio_usd'], inplace=True)

__Al ya agotar las instancias de recupero de precios, procedo a eliminar las columnas auxiliares de precios y me quedo con precio_usd__

In [40]:
df.drop(columns=['price','currency','price_aprox_local_currency','price_aprox_usd','precioReal1','regex_titulo',
                 'precioReal2','regex_desc'], inplace=True)

In [41]:
df.sample(1)

property_type   place_name  \
41444            PH  Ramos Mejía   

                                 place_with_parent_names country_name  \
41444  |Argentina|Bs.As. G.B.A. Zona Oeste|La Matanza...    Argentina   

                     state_name  geonames_id lat-lon  lat  lon  \
41444  Bs.As. G.B.A. Zona Oeste    3429617.0     NaN  NaN  NaN   

       surface_total_in_m2  surface_covered_in_m2  price_usd_per_m2  \
41444                121.0                   96.0       1239.669421   

       price_per_m2  floor  rooms  expenses  \
41444        1562.5    NaN    3.0       NaN   

                                           properati_url  \
41444  http://www.properati.com.ar/18u6j_venta_ph_ram...   

                                             description        title  \
41444  RAMOS MEJIA, CASA EN PH 3 AMBIENTES EN ALQUILE...  PH EN VENTA   

      image_thumbnail  precio_usd  
41444             NaN    150000.0

## m2 totales y cubiertos

__Chequeo a ver si hay propiedades con los m2 invertidos__

In [42]:
len(df[df.surface_total_in_m2 < df.surface_covered_in_m2])

1011

__De las filas que poseen m2 cargados, chequeo que la superficie total sea mayor que la cubierta, caso contrario las invierte, lo mismo con los precios por m2__

In [43]:
df.loc[(df.surface_total_in_m2 < df.surface_covered_in_m2),
       'sup_auxiliar'] = df.loc[(df.surface_total_in_m2 < df.surface_covered_in_m2),'surface_total_in_m2']
df.loc[(df.surface_total_in_m2 < df.surface_covered_in_m2),
       'surface_total_in_m2'] = df.loc[(df.surface_total_in_m2 < df.surface_covered_in_m2),'surface_covered_in_m2']
df.loc[(df.sup_auxiliar.notnull()),
       'surface_covered_in_m2'] = df.loc[(df.sup_auxiliar.notnull()),'sup_auxiliar']
df.loc[(df.price_usd_per_m2 > df.price_per_m2),
       'precio_aux'] = df.loc[(df.price_usd_per_m2 > df.price_per_m2),'price_usd_per_m2']
df.loc[(df.price_usd_per_m2 > df.price_per_m2),
       'price_usd_per_m2'] = df.loc[(df.price_usd_per_m2 > df.price_per_m2),'price_per_m2']
df.loc[(df.precio_aux.notnull()),
       'price_per_m2'] = df.loc[(df.precio_aux.notnull()),'precio_aux']
df.drop(columns=['sup_auxiliar','precio_aux'],inplace=True)

__Cuento la cantidad de registros sin metros cuadrados cubiertos ni totales__

In [44]:
len(df[df.surface_total_in_m2.isnull() & df.surface_covered_in_m2.isnull()])

7682

In [45]:
len(df[df.surface_covered_in_m2.isnull()])

13223

In [46]:
len(df[df.surface_total_in_m2.isnull()])

32272

__Chequeo aquellas propiedades que tengan superficie cubierta cargada pero no total__

In [47]:
len(df[df.surface_total_in_m2.isnull() & df.surface_covered_in_m2.notnull()])

24590

In [48]:
df[df.surface_total_in_m2.isnull() & df.surface_covered_in_m2.notnull()].describe().round(0)

geonames_id      lat      lon  surface_total_in_m2  \
count      20401.0  14358.0  14358.0                  0.0   
mean     3576085.0    -34.0    -59.0                  NaN   
std       378819.0      2.0      2.0                  NaN   
min      3427208.0    -41.0    -72.0                  NaN   
25%      3430234.0    -35.0    -59.0                  NaN   
50%      3433359.0    -35.0    -58.0                  NaN   
75%      3836277.0    -34.0    -58.0                  NaN   
max      6942538.0    -24.0    -55.0                  NaN   

       surface_covered_in_m2  price_usd_per_m2  price_per_m2   floor    rooms  \
count                24590.0               0.0       24308.0  3421.0  12741.0   
mean                   140.0               NaN        7726.0    33.0      3.0   
std                   1248.0               NaN       48165.0   173.0      2.0   
min                      1.0               NaN           3.0     1.0      1.0   
25%                     46.0               NaN        1412.0     1.0      2.0   
50%                     78.0               NaN        2000.0     3.0      3.0   
75%                    152.0               NaN        3040.0     6.0      4.0   
max                 187000.0               NaN     4000000.0  2509.0     30.0   

       expenses  precio_usd  
count    1807.0     24590.0  
mean     2991.0    228621.0  
std      2808.0    298416.0  
min         1.0      5000.0  
25%      1000.0     89000.0  
50%      2000.0    142000.0  
75%      4200.0    250000.0  
max     23000.0   9000000.0

__Copio los valores de superficie cubierta en la total para estos casos__

In [49]:
df.loc[(df.surface_total_in_m2.isnull() & 
        df.surface_covered_in_m2.notnull()),'surface_total_in_m2'] = df.loc[(df.surface_total_in_m2.isnull() & 
                                                       df.surface_covered_in_m2.notnull()),'surface_covered_in_m2']

Vuelvo a contar la cantidad de registros sin metros cuadrados totales

In [50]:
len(df[df.surface_total_in_m2.isnull()])

7682

Cuento la cantidad de registros sin metros cuadrados cubiertos

In [51]:
len(df[df.surface_covered_in_m2.isnull()])

13223

__Copio los valores de superficie total en la cubierta para los casos que este la total pero no la cubierta__

In [52]:
df.loc[(df.surface_total_in_m2.notnull() & 
        df.surface_covered_in_m2.isnull()),'surface_covered_in_m2'] = df.loc[(df.surface_total_in_m2.notnull() & 
                                                       df.surface_covered_in_m2.isnull()),'surface_total_in_m2']

Vuelvo a contar la cantidad de registros sin metros cuadrados totales

In [53]:
len(df[df.surface_total_in_m2.isnull()])

7682

Vuelvo a contar la cantidad de registros sin metros cuadrados cubiertos

In [54]:
len(df[df.surface_covered_in_m2.isnull()])

7682

__Para los casos restantes, intentaremos recuperar los registros con los datos de precio x m2 y el precio en USD__

In [55]:
len(df[df.surface_covered_in_m2.isnull() & (df.price_usd_per_m2.notnull() | df.price_per_m2.notnull())])

0

Como no se puede salvar por este medio, se recurre a un regex de titulo y descripción

__Comienzo por la descripcion__

In [56]:
df['regex_mts2']=df['description'].str.extract('(\d+[\.\s]?\d+\ ?[mM][tT]?[sS]?[²2])', expand=False)
len(df[df.regex_mts2.notnull() & (df.surface_covered_in_m2.isnull() | df.surface_total_in_m2.isnull())])

1111

Hay 1111 valores que pueden rescatarse con la descripción

__Sigo con el título__

In [57]:
df['regex_title_mts2']=df['title'].str.extract('(\d+[\.\s]?\d+\ ?[mM][tT]?[sS]?[²2])', expand=False)
len(df[df.regex_title_mts2.notnull() & (df.surface_covered_in_m2.isnull() | df.surface_total_in_m2.isnull())])

60

Hay 60 valores que pueden rescatarse con el titulo

In [58]:
len(df[(df.regex_mts2.notnull()) & (df.regex_title_mts2.notnull())])

2584

__Primero asigno el metraje del titulo a los m2 totales si son nulos__

In [59]:
df.loc[(df.surface_total_in_m2.isnull()) & 
       (df.surface_covered_in_m2.isnull()) &
       (df.regex_title_mts2.notnull()),'regex_final_mts2'] = df.loc[(df.surface_total_in_m2.isnull()) & 
                                                                    (df.surface_covered_in_m2.isnull()) &
                                                                    (df.regex_title_mts2.notnull()),'regex_title_mts2']

In [60]:
len(df[(df.surface_total_in_m2.isnull() & df.surface_covered_in_m2.isnull()) & (df.regex_final_mts2.notnull())])

60

Se rescataron 60 registros

__Ahora aplico lo mismo con el regex descripcion__

In [61]:
df.loc[(df.surface_total_in_m2.isnull()) & 
       (df.surface_covered_in_m2.isnull()) & 
       (df.regex_final_mts2.isnull()) & 
       (df.regex_mts2.notnull()),'regex_final_mts2'] = df.loc[(df.surface_total_in_m2.isnull()) & 
                                                              (df.surface_covered_in_m2.isnull()) & 
                                                              (df.regex_final_mts2.isnull()) & 
                                                              (df.regex_mts2.notnull()),'regex_mts2']

In [62]:
len(df[(df.surface_total_in_m2.isnull() & df.surface_covered_in_m2.isnull()) & (df.regex_final_mts2.notnull())])

1120

Se rescataron entre la descripcion y el titulo 1120 valores de m2

In [63]:
df[df.regex_final_mts2.notnull()].sample(2)

property_type   place_name  \
91498         house  Grand Bourg   
73466         house      Burzaco   

                                 place_with_parent_names country_name  \
91498  |Argentina|Bs.As. G.B.A. Zona Norte|Malvinas A...    Argentina   
73466  |Argentina|Bs.As. G.B.A. Zona Sur|Almirante Br...    Argentina   

                     state_name  geonames_id lat-lon  lat  lon  \
91498  Bs.As. G.B.A. Zona Norte          NaN     NaN  NaN  NaN   
73466    Bs.As. G.B.A. Zona Sur    3435894.0     NaN  NaN  NaN   

       surface_total_in_m2  surface_covered_in_m2  price_usd_per_m2  \
91498                  NaN                    NaN               NaN   
73466                  NaN                    NaN               NaN   

       price_per_m2  floor  rooms  expenses  \
91498           NaN    NaN    NaN       NaN   
73466           NaN    NaN    NaN       NaN   

                                           properati_url  \
91498  http://www.properati.com.ar/1bbnm_venta_casa_g...   
73466  http://www.properati.com.ar/1anqp_venta_casa_b...   

                                             description  \
91498  CASA  SOBRE LOTE 10x22 (220 M2)DOS (2) Dormito...   
73466  SOBRE CALLE ASAMBLEA Nº 1.763 (BURZACO) ENTRE ...   

                                               title  \
91498  EXCELENTE UBICACION CASA 3 Ambientes EN VENTA   
73466                                   Vende:Chalet   

                                         image_thumbnail  precio_usd  \
91498  https://thumbs4.properati.com/5/5lWVK5KPUDFCya...     80000.0   
73466  https://thumbs4.properati.com/1/laJc65r5IG0BWn...    250000.0   

      regex_mts2 regex_title_mts2 regex_final_mts2  
91498     220 M2              NaN           220 M2  
73466      130M2              NaN            130M2

Como el metraje esta escrito de varias formas, no puedo usar un solo regex, por lo que se va desde el caso más genérico y se lo pisa posteriormente con los casos más específicos para que pueda ser convertido a float

__Saco los valores raw, que se pisarán con los especiales que tienen coma__

In [64]:
df['regex_final_miles']=df['regex_final_mts2'].str.extract('(\d+[\.\s]?[0-9]?[0-9]?[0-9])', expand=False)
len(df[df.regex_final_miles.notnull()])

1120

__Saco los valores que estan en miles y que tienen coma__

In [65]:
df['regex_final_miles_coma']=df['regex_final_mts2'].str.extract('(\d+[\.\s]?[0-9][0-9][0-9][\.])', expand=False)
len(df[df.regex_final_miles_coma.notnull()])

2

__Saco los valores que estan en cientos y que tienen coma__

In [66]:
df['regex_final_cientos_coma']=df['regex_final_mts2'].str.extract('(\d+[\.\s]?\d+[\.])', expand=False)
len(df[df.regex_final_miles_coma.isnull() & df.regex_final_cientos_coma.notnull()])

48

__Paso a float la columna con los valores raw sin las excepciones que tienen coma__

In [67]:
df.regex_final_miles.replace(to_replace='\.',value='',inplace=True, regex=True)
df.regex_final_miles.replace(to_replace='\s',value='.',inplace=True, regex=True)
df.regex_final_miles = df.regex_final_miles.astype(float)
len(df[df.regex_final_miles.notnull()])

1120

Chequeo conversión a float

In [68]:
df.regex_final_miles.dtype

dtype('float64')

__Paso la columna de miles con coma a float__

In [69]:
df.regex_final_miles_coma = df.regex_final_miles_coma.astype(float)
df.regex_final_miles_coma.dtype

dtype('float64')

__Paso la columna de cientos con coma a float__

In [70]:
df.regex_final_cientos_coma = df.regex_final_cientos_coma.astype(float)
df.regex_final_cientos_coma.dtype

dtype('float64')

__Tomo la columna de miles y la voy pisando con los valores de cientos y miles con coma__

In [71]:
df.loc[df.regex_final_cientos_coma.notnull(),
       'regex_final_miles'] = df.loc[df.regex_final_cientos_coma.notnull(), 'regex_final_cientos_coma']
df.loc[df.regex_final_miles_coma.notnull(),
       'regex_final_miles'] = df.loc[df.regex_final_miles_coma.notnull(), 'regex_final_miles_coma']

__Como ya no tengo más valores de m2 para salvar, asigno los valores finales de regex_final_miles a surface_total_in_m2 y a surface_covered_in_m2__

In [72]:
df.loc[(df.surface_total_in_m2.isnull()) & 
       (df.surface_covered_in_m2.isnull()) & 
       (df.regex_final_miles.notnull()),
       ['surface_total_in_m2','surface_covered_in_m2']] = df.loc[(df.surface_total_in_m2.isnull()) & 
                                                                   (df.surface_covered_in_m2.isnull()) & 
                                                                   (df.regex_final_miles.notnull()),'regex_final_miles']

Procedo a eliminar las columnas auxiliares de m2

In [73]:
df.drop(columns=['regex_mts2','regex_title_mts2','regex_final_mts2','regex_final_miles','regex_final_miles_coma',
                 'regex_final_cientos_coma'], inplace=True)

In [74]:
df.isnull().sum()

property_type                  0
place_name                    23
place_with_parent_names        0
country_name                   0
state_name                     0
geonames_id                17752
lat-lon                    42389
lat                        42389
lon                        42389
surface_total_in_m2         6562
surface_covered_in_m2       6562
price_usd_per_m2           34610
price_per_m2               15569
floor                      96235
rooms                      61401
expenses                   89978
properati_url                  0
description                    2
title                          0
image_thumbnail             2252
precio_usd                     0
dtype: int64

Elimino las filas que no poseen metros

In [75]:
df.dropna(subset=['surface_total_in_m2'], inplace=True)

## Precio por m2 total y cubierto

Redondeo los valores de precio por m2

In [76]:
df['price_usd_per_m2'] = df.price_usd_per_m2.round(0)
df['price_per_m2'] = df.price_per_m2.round(0)

__Ahora con los valores de m2 cubiertos y no cubiertos, procedemos al cálculo del precio por m2 para todos los valores__

In [77]:
df['precio_m2_total'] = (df.precio_usd / df.surface_total_in_m2)
df['precio_m2_cubierto'] = df.precio_usd / df.surface_covered_in_m2
df['precio_m2_total'] = df.precio_m2_total.round(0)
df['precio_m2_cubierto'] = df.precio_m2_cubierto.round(0)
df.sample(1)

property_type place_name      place_with_parent_names country_name  \
66513     apartment    Córdoba  |Argentina|Córdoba|Córdoba|    Argentina   

      state_name  geonames_id lat-lon  lat  lon  surface_total_in_m2  \
66513    Córdoba    3860259.0     NaN  NaN  NaN                 53.0   

       surface_covered_in_m2  price_usd_per_m2  price_per_m2  floor  rooms  \
66513                   53.0            1051.0           NaN    NaN    NaN   

       expenses                                      properati_url  \
66513       NaN  http://www.properati.com.ar/1adww_venta_depart...   

                                             description  \
66513  Te invitamos a conocer: Octavio VI Característ...   

                             title  \
66513  Departamento en General Paz   

                                         image_thumbnail  precio_usd  \
66513  https://thumbs4.properati.com/5/iP64aMvQrVua61...    55718.01   

       precio_m2_total  precio_m2_cubierto  
66513           1051.0              1051.0

__Calculo diferencias de precios con los ya dados, para ver outliers__

In [78]:
df['dif_precio_m2_total'] = df.price_usd_per_m2 - df.precio_m2_total
df['dif_precio_m2_cubierto'] = df.price_per_m2 - df.precio_m2_cubierto

In [79]:
df.dif_precio_m2_total.describe().round(2)

count    68617.00
mean         0.98
std         56.04
min          0.00
25%          0.00
50%          0.00
75%          0.00
max       9121.00
Name: dif_precio_m2_total, dtype: float64

In [80]:
len(df[df.dif_precio_m2_total > 1])

88

Solo 88 valores estan por encima de diferencias de 1 dolar, lo que equivale a menos del 1%

__Chequeo que el precio por m2 total no este intercambiado con el precio por m2 cubierto__

In [81]:
df['precio_m2_intercambiado'] = df.price_usd_per_m2 - df.precio_m2_cubierto

In [82]:
len(df[(df.dif_precio_m2_total > 1) & (df.precio_m2_intercambiado == 0)])

83

Como se puede ver 83 valores de los 88 estan intercambiados, por lo que se conserva estos valores, y los restantes 5 se los elimina

In [83]:
df.drop(axis=0, index=df.index[((df.dif_precio_m2_total > 1) & (df.precio_m2_intercambiado != 0))], inplace=True)

__Chequeo que se hayan eliminado__

In [84]:
len(df.index[((df.dif_precio_m2_total > 1) & (df.precio_m2_intercambiado != 0))])

0

__Elimino la columna auxiliar 'precio_m2_intercambiado' y 'dif_precio_m2_total'__

In [85]:
df.drop(columns=['precio_m2_intercambiado','dif_precio_m2_total'], inplace=True)

__Procedo a revisar la diferencia en los m2 cubiertos__

In [86]:
df.dif_precio_m2_cubierto.describe().round(2)

count      87653.00
mean        3749.95
std        20784.64
min            0.00
25%            0.00
50%            0.00
75%            0.00
max      3775665.00
Name: dif_precio_m2_cubierto, dtype: float64

In [87]:
len(df[df.dif_precio_m2_cubierto > 1])

10423

En este caso se pueden apreciar muchas filas con diferencias. En una vista rápida del dataframe se puede ver que los valores de price_per_m2 son inusualmente altos vs los calculados, por lo que puede ser que el valor este en AR$. Se procede a aplicarle el tc y se calculan las diferencias nuevamente.

In [88]:
df['precio_m2_dolarizado']=df.price_per_m2/tc

In [89]:
df['dif_precio_m2_cubierto_dolarizado'] = df.precio_m2_cubierto - df.precio_m2_dolarizado

In [90]:
df[(df.dif_precio_m2_cubierto > 1)]['dif_precio_m2_cubierto_dolarizado'].describe().round(2)

count    10423.00
mean       -10.94
std        148.36
min      -2364.54
25%        -22.80
50%        -17.48
75%        -11.50
max       9078.08
Name: dif_precio_m2_cubierto_dolarizado, dtype: float64

Ahora se puede ver que las diferencias se acotan en números más cercanos al valor calculado.

__Chequeo que cantidad de propiedades estan dentro de un rango "razonable" al calculado. (Se toma la cota de una diferencia de -50 y +50 dolares/m2__

In [91]:
len(df[(df.dif_precio_m2_cubierto > 1) & (df.dif_precio_m2_cubierto_dolarizado > -50) &(df.dif_precio_m2_cubierto_dolarizado < 50)])

10044

La cantidad es casi la misma que tenía diferencias altas previamente, por lo que se decide mantener los valores calculados para todos los casos. Se chequeará posteriormente por outliers para los restantes casos

__Al ya tener todos los valores de precio por m2, procedo a eliminar las columnas auxiliares__

In [92]:
df.drop(columns=['price_usd_per_m2','price_per_m2','dif_precio_m2_cubierto',
                 'precio_m2_dolarizado','dif_precio_m2_cubierto_dolarizado'], inplace=True)

## Ubicación

Tomo la columna 'place_with_parent_names' y voy separando los conceptos por '|'. En caso que sea el nombre de un barrio privado se agregó la columna extra_info.

In [93]:
df['pais']=df["place_with_parent_names"].str.split('|').str[1]
df['provincia']=df["place_with_parent_names"].str.split('|').str[2]
df['barrio_ciudad']=df["place_with_parent_names"].str.split('|').str[3]
df['extra_info']=df["place_with_parent_names"].str.split('|').str[4]
df['extra_info_2']=df["place_with_parent_names"].str.split('|').str[5]

__Procedo a reemplazar los espacios en blanco por NaN's__

In [94]:
len(df[(df.pais == '') | (df.pais.isnull())])

0

In [95]:
len(df[(df.provincia == '') | (df.provincia.isnull())])

0

In [96]:
len(df[(df.barrio_ciudad == '') | (df.barrio_ciudad.isnull())])

3241

In [97]:
df.loc[df.barrio_ciudad == '','barrio_ciudad'] = np.nan
len(df[(df.barrio_ciudad == '')])

0

In [98]:
len(df[(df.extra_info == '') | (df.extra_info.isnull())])

61697

In [99]:
df.loc[df.extra_info == '','extra_info'] = np.nan
len(df[(df.extra_info == '')])

0

In [100]:
len(df[(df.extra_info_2 == '') | (df.extra_info_2.isnull())])

96120

In [101]:
df.loc[df.extra_info_2 == '','extra_info_2'] = np.nan
len(df[(df.extra_info_2 == '')])

0

In [102]:
len(df)

96660

__Chequeo que el pais y la provincia estén bien__

In [103]:
len(df[(df.country_name == df.pais) == False])

0

In [104]:
len(df[(df.state_name == df.provincia) == False])

0

Como las columnas son iguales, elimino las columnas country_name y state_name

In [105]:
df.drop(columns=['country_name','state_name'],inplace=True)

__Chequeo que los valores de extra_info_2 coincidan con los de place_name__

In [106]:
len(df[df.extra_info_2 == df.place_name])

540

Como se puede ver, el place_name de los valores de extra_info_2 están bien.

__Chequeo que los valores de extra_info coincidan con los de place_name__

In [107]:
len(df[df.extra_info.notnull() & df.extra_info_2.isnull()])

34423

In [108]:
len(df[(df.extra_info == df.place_name) & (df.extra_info.notnull()) & (df.extra_info_2.isnull())])

34423

Como se puede ver, el place_name de los valores de extra_info están bien.

__Chequeo que los valores de barrio_ciudad coincidan con los de place_name__

In [109]:
len(df[df.barrio_ciudad.notnull() & df.extra_info.isnull() & df.extra_info_2.isnull()])

58456

In [110]:
len(df[(df.barrio_ciudad == df.place_name) & (df.barrio_ciudad.notnull())
       & (df.extra_info.isnull()) & (df.extra_info_2.isnull())])

58433

Como se puede ver, el place_name de los valores de barrio_ciudad están bien excepto por 23 valores

In [111]:
len(df[(df.place_name.isnull()) & (df.barrio_ciudad.notnull()) 
       & (df.extra_info.isnull()) & (df.extra_info_2.isnull())])

23

Los 23 valores de place_name que son null tienen su valor en barrio_ciudad, se lo asigno a place_name

In [112]:
df.loc[(df.place_name.isnull()) & (df.barrio_ciudad.notnull()) & (df.extra_info.isnull()) 
       & (df.extra_info_2.isnull()),'place_name'] = df.loc[(df.place_name.isnull()) & (df.barrio_ciudad.notnull()) 
                                                           & (df.extra_info.isnull()) & (df.extra_info_2.isnull()),
                                                           'barrio_ciudad']
len(df[(df.place_name.isnull()) & (df.barrio_ciudad.notnull()) 
       & (df.extra_info.isnull()) & (df.extra_info_2.isnull())])

0

Ahora que barrio_ciudad coincide con place_name prosigo con la provincia

__Chequeo que los valores de provincia coincidan con los de place_name__

In [113]:
len(df[df.provincia.notnull() & df.barrio_ciudad.isnull() & df.extra_info.isnull() & df.extra_info_2.isnull()])

3241

In [114]:
len(df[(df.provincia == df.place_name) & df.provincia.notnull() & df.barrio_ciudad.isnull() 
       & df.extra_info.isnull() & df.extra_info_2.isnull()])

3241

Como se puede ver, el place_name de los valores de provincia están bien.

__Chequeo que los valores de pais coincidan con los de place_name__

In [115]:
len(df[df.pais.notnull() & df.provincia.isnull() & df.barrio_ciudad.isnull() 
       & df.extra_info.isnull() & df.extra_info_2.isnull()])

0

No hay place_name que sean solo país

In [116]:
len(df.place_name.notnull())-len(df[df.extra_info_2 
                                    == df.place_name])-len(df[(df.extra_info == df.place_name) 
                                                              & (df.extra_info.notnull()) 
                                                              & (df.extra_info_2.isnull())]) - len(df[(df.barrio_ciudad == df.place_name) 
                                                                                                      & (df.barrio_ciudad.notnull()) 
                                                                                                      & (df.extra_info.isnull()) 
                                                                                                      & (df.extra_info_2.isnull())]) - len(df[(df.provincia == df.place_name) 
                                                                                                                                              & df.provincia.notnull() 
                                                                                                                                              & df.barrio_ciudad.isnull() 
                                                                                                                                              & df.extra_info.isnull() 
                                                                                                                                              & df.extra_info_2.isnull()])

0

Se verifica que para todos los place_name tiene asignado una clasificación, con lo cual procedo a eliminar las columnas que no son necesarias.

In [117]:
df.drop(columns=['place_name','place_with_parent_names'], inplace=True)

__Chequeo valores de la columna Pais__

In [118]:
df.pais.value_counts()

Argentina    96660
Name: pais, dtype: int64

Como todas las propiedades son de argentina no tiene sentido guardar esta información en el dataset, se elimina

In [119]:
df.drop(columns=['pais'], inplace=True)

__Ordeno el dataframe con los datos finales__

In [120]:
list(df.columns.values)

['property_type',
 'geonames_id',
 'lat-lon',
 'lat',
 'lon',
 'surface_total_in_m2',
 'surface_covered_in_m2',
 'floor',
 'rooms',
 'expenses',
 'properati_url',
 'description',
 'title',
 'image_thumbnail',
 'precio_usd',
 'precio_m2_total',
 'precio_m2_cubierto',
 'provincia',
 'barrio_ciudad',
 'extra_info',
 'extra_info_2']

In [121]:
df = df[['precio_usd','provincia','barrio_ciudad','extra_info','extra_info_2','property_type','surface_total_in_m2',
         'surface_covered_in_m2','precio_m2_total','precio_m2_cubierto','geonames_id','lat-lon','lat','lon','floor',
         'rooms','expenses','properati_url','description','title','image_thumbnail']]
df.shape

(96660, 21)

In [122]:
df.sample(1)

precio_usd        provincia  barrio_ciudad extra_info extra_info_2  \
93962    550000.0  Capital Federal  Puerto Madero        NaN          NaN   

      property_type  surface_total_in_m2  surface_covered_in_m2  \
93962     apartment                109.0                   96.0   

       precio_m2_total  precio_m2_cubierto  geonames_id lat-lon  lat  lon  \
93962           5046.0              5729.0    3429760.0     NaN  NaN  NaN   

       floor  rooms  expenses  \
93962    NaN    NaN    8400.0   

                                           properati_url  \
93962  http://www.properati.com.ar/1bevv_venta_depart...   

                                             description  \
93962  Terrazas del Dique. Excelente 4 ambientes con ...   

                                      title  \
93962  TERRAZAS  DEL DIQUE 4 AMB.   COCHERA   

                                         image_thumbnail  
93962  https://thumbs4.properati.com/0/FXkhFJi4Kss-hi...

# SEGUIR ACA CON TIPO DE PROPIEDAD

## Piso

__Arranco revisando la variable__

In [123]:
df.floor.describe().round(2)

count    6795.00
mean       18.89
std       123.51
min         1.00
25%         1.00
50%         3.00
75%         6.00
max      2509.00
Name: floor, dtype: float64

Como las torres más altas de la argentina no tienen más de 50 pisos y los de más de 35 pisos se encuentran en Puerto Madero, se setea con NaN los valores mayores a estos parámetros.

In [124]:
df.loc[((df.floor > 35) & (df.barrio_ciudad != 'Puerto Madero')),'floor'] = np.nan

# Análisis descriptivo de variables

__Procedo a la selección de variables que no tengan nulos en el df__

In [125]:
df.isnull().sum()

precio_usd                   0
provincia                    0
barrio_ciudad             3241
extra_info               61697
extra_info_2             96120
property_type                0
surface_total_in_m2          0
surface_covered_in_m2        0
precio_m2_total              0
precio_m2_cubierto           0
geonames_id              16862
lat-lon                  39522
lat                      39522
lon                      39522
floor                    90057
rooms                    56575
expenses                 83752
properati_url                0
description                  1
title                        0
image_thumbnail           1897
dtype: int64

In [126]:
dfg = df.drop(columns=['extra_info','geonames_id','lat-lon','lat','lon','floor','expenses',
                       'rooms','properati_url','description','title','image_thumbnail'])
dfg.sample(1)

precio_usd                 provincia barrio_ciudad extra_info_2  \
12567    182000.0  Bs.As. G.B.A. Zona Norte  San Fernando          NaN   

      property_type  surface_total_in_m2  surface_covered_in_m2  \
12567     apartment                 59.0                   49.0   

       precio_m2_total  precio_m2_cubierto  
12567           3085.0              3714.0

In [127]:
df[df.surface_total_in_m2>=df.precio_usd]

precio_usd                     provincia  barrio_ciudad extra_info  \
11534     49800.0  Buenos Aires Costa Atlántica  Mar del Plata        NaN   
13405    120000.0                      Santa Fe            NaN        NaN   

      extra_info_2 property_type  surface_total_in_m2  surface_covered_in_m2  \
11534          NaN     apartment              49800.0                49800.0   
13405          NaN     apartment             200000.0                   80.0   

       precio_m2_total  precio_m2_cubierto  geonames_id  \
11534              1.0                 1.0    3430863.0   
13405              1.0              1500.0    3836276.0   

                       lat-lon        lat        lon  floor  rooms  expenses  \
11534  -38.0051676,-57.5444832 -38.005168 -57.544483    NaN    1.0       NaN   
13405  -31.6400864,-60.7046147 -31.640086 -60.704615    NaN    3.0    1000.0   

                                           properati_url  \
11534  http://www.properati.com.ar/16bzw_venta_depart...   
13405  http://www.properati.com.ar/16p2j_venta_depart...   

                                             description  \
11534  TOTALMENTE LUMINOSO LIVING COMEDOR CON AMPLIO ...   
13405  ambientes amplios, iluminados a la calle, idea...   

                       title  \
11534  DEPARTAMENTO EN VENTA   
13405     Departamento venta   

                                         image_thumbnail  
11534  https://thumbs4.properati.com/6/vggfdgnqIgFMN4...  
13405  https://thumbs4.properati.com/3/x6imA3e2uBBgO3...

In [128]:
df=df.drop([11534])

In [129]:
df.loc[13405,'surface_total_in_m2']= df.loc[13405,'surface_covered_in_m2']

In [130]:
df2=df[['title','description','lat','lon','precio_usd','barrio_ciudad','property_type','surface_total_in_m2']]
df.drop(df2[df2.duplicated()].index, axis=0, inplace=True)
len(df)

94116

In [131]:
barrios=df.barrio_ciudad.unique()
for x in barrios:
        mean_1 = np.mean(df.precio_usd[df.barrio_ciudad==x])
        std_1 =np.std(df.precio_usd[df.barrio_ciudad==x])

        for index2, y in df[df.barrio_ciudad==x].iterrows():
                if(std_1!=0):
                    z_score= (y.precio_usd - mean_1)/std_1 
                    if np.abs(z_score) > 2.5:
                        df.loc[index2,'outlier']="YES"

In [132]:
len(df[df.outlier=="YES"])

2114

In [133]:
df.drop(df[df.outlier=="YES"].index, axis=0, inplace=True)

In [134]:
len(df)

92002

In [135]:
# df.loc[df.cero == 'NO', 'zscore'] =  df.loc[df.cero == 'NO', 'precio'] +

In [136]:
import math 
df.drop(df[df.precio_m2_total==math.inf].index, axis=0, inplace=True)

In [137]:
scoreProvincia=[]
provincia=df.provincia.unique()
provincia=provincia.tolist()
del(provincia[27])
provincia

['Capital Federal',
 'Buenos Aires Costa Atlántica',
 'Bs.As. G.B.A. Zona Norte',
 'Córdoba',
 'Bs.As. G.B.A. Zona Oeste',
 'Bs.As. G.B.A. Zona Sur',
 'Santa Fe',
 'Misiones',
 'Entre Ríos',
 'Buenos Aires Interior',
 'Neuquén',
 'Río Negro',
 'Mendoza',
 'Corrientes',
 'Chubut',
 'Salta',
 'San Luis',
 'Tucumán',
 'Tierra Del Fuego',
 'Chaco',
 'Catamarca',
 'Santa Cruz',
 'La Pampa',
 'Santiago Del Estero',
 'Jujuy',
 'San Juan',
 'La Rioja']

In [138]:
def scoreTrainProvincia(prov):

    df_capFed=df.loc[df['provincia']==prov]


    df_capFed=df_capFed[['barrio_ciudad','property_type','surface_total_in_m2','surface_covered_in_m2','precio_m2_total']]


    df_capFed = df_capFed.reset_index()

    df_capFed['precio_m2_total'] = np.log(df_capFed['precio_m2_total'])

    df_capFed['precio_m2_total_new'] = pd.Series(len(df_capFed['precio_m2_total']), index=df_capFed.index)
    df_capFed['precio_m2_total_new'] = 0 
    df_capFed.loc[df_capFed['precio_m2_total']>0,'precio_m2_total_new'] = 1

    df_capFed.loc[df_capFed['precio_m2_total_new']==1,'precio_m2_total'] = np.log(df_capFed['precio_m2_total'])

    numeric_feats = df_capFed.dtypes[df_capFed.dtypes != "object"].index

    # Check the skew of all numerical features
    skewed_feats = df_capFed[numeric_feats].apply(lambda x: skew(x.dropna())).sort_values(ascending=False)
    skewness = pd.DataFrame({'Skew' :skewed_feats})


    skewness = skewness[abs(skewness) > 0.75]

    skewed_features = skewness.index
    lam = 0.15
    for feat in skewed_features:
        #all_data[feat] += 1
        df_capFed[feat] = boxcox1p(df_capFed[feat], lam)

    #df_capFed=df_capFed.dropna()

    #df_capFed.head()

    df_capFed=pd.get_dummies(df_capFed, drop_first=True)

    y1=df_capFed.precio_m2_total


    df_capFed=df_capFed.drop('precio_m2_total', axis=1)

    X1= PolynomialFeatures(2,include_bias=False,interaction_only=False).fit_transform(df_capFed) 
    
    feature=list(X1.shape)
    feature=feature[1]
    component=math.trunc(feature)
    if component<5:
        unicaPropiedad.append(prov)
        return


    X_train, X_test, y_train, y_test = train_test_split(X1, y1, test_size=0.30, random_state=7)


    scaler.fit(X_train)

    scaler.transform(X_train)
    scaler.transform(X_test)

    pca_1cp = PCA(svd_solver='auto')
    pca_1cp.fit(X_train)

    X_pca = pca_1cp.transform(X_train)
    X_pca = pca_1cp.transform(X_test)


    al_ridge = np.linspace(0.001, 0.3, 100)#buscar que hace linspace 
    al_lasso = np.linspace(0.1, 0.5, 100)
    kf = KFold(n_splits=5, shuffle=True, random_state=12)#buscar que es kfold 

    # Instanciamos los modelos

    lm = LinearRegression()

    lm_ridge_cv= RidgeCV(alphas=al_ridge, cv=kf, normalize=False)#hace el cross validation automatico 

    lm_lasso_cv = LassoCV(alphas=al_lasso, cv=kf, normalize=False)

    GBoost = GradientBoostingRegressor(n_estimators=3000, learning_rate=0.05,
                                       max_depth=4, max_features='sqrt',
                                       min_samples_leaf=15, min_samples_split=10, 
                                       loss='huber', random_state =5)

    model_lgb = lgb.LGBMRegressor(objective='regression',num_leaves=50000,
                                  learning_rate=0.05, n_estimators=720,
                                  max_bin = 55, bagging_fraction = 0.8,
                                  bagging_freq = 5, feature_fraction = 0.2319,
                                  feature_fraction_seed=9, bagging_seed=9,
                                  min_data_in_leaf =6, min_sum_hessian_in_leaf = 11)

    regr_1 = DecisionTreeRegressor(max_depth=1000)


    models=[lm_ridge_cv,
            GBoost,
            model_lgb,
           regr_1]
           #lm]

    S_train, S_test = stacking(models,                     # list of models
                               X_train, y_train, X_test,   # data
                               regression=True,            # regression task (if you need 
                                                           #     classification - set to False)
                               mode='oof_pred_bag',        # mode: oof for train set, predict test 
                                                           #     set in each fold and find mean
                               save_dir=None,              # do not save result and log (to save 
                                                           #     in current dir - set to '.')
                                                           # metric: callable
                               n_folds=4,                  # number of folds
                               shuffle=True,               # shuffle the data
                               random_state=0,             # ensure reproducibility
                               verbose=0)                  # print all info

    rmse = lambda y, y_pred: np.sqrt(mean_squared_error(y, y_pred))

    # Initialize 2nd level model

    model = XGBRegressor(random_state=0, n_jobs=-1, learning_rate=0.1, 
                         n_estimators=70, max_depth=7)

    # Fit 2nd level model
    model = model.fit(S_train, y_train)

    # Predict
    y_pred = model.predict(S_test)

    scoreProvincia.append({"provincia":prov,"score":metrics.r2_score(y_test, y_pred),"rmse":np.sqrt(metrics.mean_squared_error(y_test, y_pred))})
    # Graficamos el modelo
    test = go.Scatter3d(
    x=y_test,
    y=y_test,
    z=y_test,
    mode='markers',
    marker=dict(
        size=8,
        line=dict(
            color='rgba(217, 217, 217, 0.14)',
            width=0.5
        ),
        opacity=1
    )
)


    pred = go.Scatter3d(
    x=y_test,
    y=y_pred,
    z=y_test,
    mode='markers',
    marker=dict(
        color='rgb(0, 255, 0)',
        size=6,
        symbol='cross',
        line=dict(
            color='rgb(204, 204, 204)',
            width=1
        ),
        opacity=0.8
    )
)
    data = [test, pred]
    layout = go.Layout(
    margin=dict(
        l=0,
        r=0,
        b=0,
        t=0
    )
)
    fig = go.Figure(data=data, layout=layout)
    
    print ('MAE:', metrics.mean_absolute_error(y_test, y_pred))
    print ('MSE:', metrics.mean_squared_error(y_test, y_pred))
    print ('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))
    print ('R2:', metrics.r2_score(y_test, y_pred))
    return py.iplot(fig, filename='simple-3d-scatter')

In [139]:
unicaPropiedad=[]
scoreProvincia=[]
for prov in provincia:
    print("-"*100)
    print(prov)
    scoreTrainProvincia(prov)

----------------------------------------------------------------------------------------------------
Capital Federal
MAE: 0.009722526527003086
MSE: 0.0002733941859079226
RMSE: 0.016534635947244882
R2: 0.6611444962248587
----------------------------------------------------------------------------------------------------
Buenos Aires Costa Atlántica
MAE: 0.013443681451641388
MSE: 0.0005222086454933061
RMSE: 0.02285188494398889
R2: 0.8484324130945013
----------------------------------------------------------------------------------------------------
Bs.As. G.B.A. Zona Norte
MAE: 0.012447819944565147
MSE: 0.0004180208385168374
RMSE: 0.020445557916497104
R2: 0.8059671939315615
----------------------------------------------------------------------------------------------------
Córdoba
MAE: 0.02130117665277236
MSE: 0.0012213047865775068
RMSE: 0.03494717136732967
R2: 0.6475806001923821
----------------------------------------------------------------------------------------------------
Bs.As. G

ValueError: Cannot have number of splits n_splits=4 greater than the number of samples: n_samples=2.

In [140]:
dfScore=pd.DataFrame(scoreProvincia)
dfScore.score=dfScore.score.astype(float)
dfScore.score=dfScore.score * 100
dfScore.score=dfScore.score.astype(int)
dfScore

provincia      rmse  score
0                Capital Federal  0.016535     66
1   Buenos Aires Costa Atlántica  0.022852     84
2       Bs.As. G.B.A. Zona Norte  0.020446     80
3                        Córdoba  0.034947     64
4       Bs.As. G.B.A. Zona Oeste  0.026643     73
5         Bs.As. G.B.A. Zona Sur  0.029652     64
6                       Santa Fe  0.038640     53
7                       Misiones  0.046383     53
8                     Entre Ríos  0.043814     40
9          Buenos Aires Interior  0.039690     82
10                       Neuquén  0.062793     31
11                     Río Negro  0.038091     60
12                       Mendoza  0.058810     68
13                    Corrientes  0.019831     63
14                        Chubut  0.041337      7
15                         Salta  0.050703    -18
16                      San Luis  0.050971     33
17                       Tucumán  0.057144    -13
18              Tierra Del Fuego  0.033625     50
19                         Chaco  0.036347     12
20                     Catamarca  0.038071     61
21                    Santa Cruz  0.055091   -119
22                      La Pampa  0.058558   -164

In [ ]:
scoreTrainProvincia("Santa Cruz")